systemtest/cmd/sendotlp/main.go

```diff
@@ -20,70 +20,147 @@ package main
+	"go.opentelemetry.io/otel/sdk/trace/tracetest"
 	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.uber.org/zap/zapgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/grpclog"
 )
 
+var (
+	endpoint = flag.String(
+		"endpoint",
+		getenvDefault("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:8200"),
+		"target URL to which sendotlp will send spans and metrics ($OTEL_EXPORTER_OTLP_ENDPOINT)",
+	)
+
+	secretToken = flag.String(
+		"secret-token",
+		"",
+		"Elastic APM secret token. Note: setting this overrides $OTEL_EXPORTER_OTLP_HEADERS",
+	)
+
+	logLevel = zap.LevelFlag(
+		"loglevel", zapcore.InfoLevel,
+		"set log level to one of DEBUG, INFO (default), WARN, ERROR, DPANIC, PANIC, FATAL",
+	)
+)
+
+func getenvDefault(key, defaultVal string) string {
+	val := os.Getenv(key)
+	if val != "" {
+		return val
+	}
+	return defaultVal
+}
+
 func main() {
-	if err := Main(context.Background()); err != nil {
-		log.Fatal(err)
+	flag.Parse()
+	zapcfg := zap.NewProductionConfig()
+	zapcfg.EncoderConfig.EncodeTime = zapcore.RFC3339TimeEncoder
+	zapcfg.EncoderConfig.EncodeLevel = zapcore.CapitalColorLevelEncoder
+	zapcfg.Encoding = "console"
+	zapcfg.Level = zap.NewAtomicLevelAt(*logLevel)
+	logger, err := zapcfg.Build()
+	if err != nil {
+		panic(err)
+	}
+	defer logger.Sync()
+	grpclog.SetLogger(zapgrpc.NewLogger(logger, zapgrpc.WithDebug()))
+
+	if err := Main(context.Background(), logger.Sugar()); err != nil {
+		logger.Fatal("error sending data", zap.Error(err))
 	}
 }
 
-func Main(ctx context.Context) error {
-	const otelEndpointEnv = "OTEL_EXPORTER_OTLP_ENDPOINT"
-	endpoint := os.Getenv(otelEndpointEnv)
-	if endpoint == "" {
-		endpoint = "http://localhost:8200"
-		os.Setenv(otelEndpointEnv, endpoint)
+func Main(ctx context.Context, logger *zap.SugaredLogger) error {
+	var insecure bool
+	endpointURL, err := url.Parse(*endpoint)
+	if err != nil {
+		return fmt.Errorf("failed to parse endpoint: %w", err)
+	}
+	switch endpointURL.Scheme {
+	case "http":
+		if endpointURL.Port() == "" {
+			endpointURL.Host = net.JoinHostPort(endpointURL.Host, "80")
+		}
+		insecure = true
+	case "https":
+		if endpointURL.Port() == "" {
+			endpointURL.Host = net.JoinHostPort(endpointURL.Host, "443")
+		}
+	default:
+		return fmt.Errorf(
+			"endpoint must be prefixed with http:// or https://",
+		)
 	}
-	log.Printf("Sending OTLP data to %s ($%s)", endpoint, otelEndpointEnv)
+	logger.Infof("sending OTLP data to %s", endpointURL.String())
+	grpcEndpoint := endpointURL.Host
 
-	var otlpTraceExporterOptions []otlptracegrpc.Option
-	otlpTraceExporterOptions = append(otlpTraceExporterOptions, otlptracegrpc.WithInsecure())
-	otlpTraceExporter, err := otlptracegrpc.New(ctx, otlpTraceExporterOptions...)
+	var grpcDialOptions []grpc.DialOption
+	if insecure {
+		// If http:// is specified, then use insecure (plaintext).
+		grpcDialOptions = append(grpcDialOptions, grpc.WithInsecure())
```

Comment:
I think `grpc.WithInsecure` has been deprecated (https://pkg.go.dev/google.golang.org/grpc#WithInsecure). in favor of `grpc.WithTransportCredentials(insecure.NewCredentials())`.

Reply:
TIL - thanks, will update.